### PR TITLE
Binder and fixes

### PR DESCRIPTION
**Pull request by MSeal**
Added a more interactive binder based on papermill's binder.

Found a bug with reglue picking up incorrect data, which caused some additional metadata additions and a rewrite of reglue to reuse existing glue code where it can.

Fixed scraps_report to be able to report on data scraps and to not persist data as associated with the parent notebook when reporting.

Last PR before 0.2.0 release.

### CODE REVIEW

File: scrapbook/tests/test_scrapbooks.py [hunk 1]
```diff
@@ -254,64 +278,85 @@ def test_scraps_report(mock_display, notebook_collection):
...
 
 
 @mock.patch("scrapbook.models.ip_display")
-def test_scraps_report_no_header(mock_display, notebook_collection):
-    notebook_collection.scraps_report(header=None)
+def test_scraps_report_no_headers(mock_display, notebook_collection):
+    notebook_collection.scraps_report(headers=None)
     mock_display.assert_has_calls(
         [
-            mock.call(
```
Comment (MSeal) [inline scrapbook/tests/test_scrapbooks.py]: black compacted all of these when the metadata became an empty dict

File: binder/requirements.txt [hunk 2]
```diff
@@ -1,2 +1,5 @@
 ../scrapbook
```
Comment (betatim) [inline binder/requirements.txt]: mybinder.org complains about this not being a valid requirement. Should it be `-e../scrapbook`?

Comment (MSeal) [reply binder/requirements.txt]: Hmm nope, tried that too. And a few other options (this is affecting papermill binders too). @willingc said she might be able to help here.

Comment (betatim) [reply binder/requirements.txt]: Ah! `-e..` as it should be the directory that contains the `setup.py`

Comment (MSeal) [reply binder/requirements.txt]: Nope, that fails too :/

Comment (MSeal) [reply binder/requirements.txt]: Carol mentioned that it might be related to how the docker image sets up such that it's not available.

Comment (willingc) [reply binder/requirements.txt]: @MSeal I applied your PR on my fork and got it to work on binder. I removed `../scrapbook` on line 1 and added a final requirement `-e .` See https://github.com/willingc/scrapbook/blob/binderAndFixes/binder/requirements.txt#L5

File: binder/reglue_highlight_dates.ipynb [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    "out_one.reglue('highlight_dates_fig')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Again, we can also explore the output notebook:\n",
+    "\n",
+    "<a href=\"highlight_dates_run_one.ipynb\"><button type=\"button\">Link to output notebook!</button></a>\n"
```
Comment (willingc) [inline binder/reglue_highlight_dates.ipynb]: This 404s as is on binder. I think it should be `<a href="./outcomes/highlight_dates_run_one.ipynb"><button type="button">Link to output notebook!</button></a>`

File: binder/reglue_highlight_dates.ipynb [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    "out_two.reglue('highlight_dates_fig')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For the second notebook source:\n",
+    "\n",
+    "<a href=\"highlight_dates_run_two.ipynb\"><button type=\"button\">Link to output notebook!</button></a>"
```
Comment (willingc) [inline binder/reglue_highlight_dates.ipynb]: Also 404. Perhaps `<a href="./outcomes/highlight_dates_run_two.ipynb"><button type="button">Link to output notebook!</button></a>`

File: scrapbook/api.py [hunk 5]
```diff
@@ -10,12 +10,13 @@
 import IPython
 
 from six import string_types
-from IPython.display import display as ip_display
+
+# from IPython.display import display as ip_display
```
Comment (willingc) [inline scrapbook/api.py]: Remove this line?

File: scrapbook/api.py [hunk 6]
```diff
@@ -84,30 +85,48 @@ def glue(name, scrap, encoder=None, display=None):
...
-        display_kwargs = {}
-        if isinstance(display, (list, tuple)):
-            display_kwargs = {"include": display}
-        elif isinstance(display, dict):
-            display_kwargs = display
-        data, metadata = IPython.core.formatters.format_display_data(
-            scrap, **display_kwargs
-        )
-        metadata["scrapbook"] = dict(name=name)
+        data, metadata = _display_scrap_display(name, scrap, display)
```
Comment (willingc) [inline scrapbook/api.py]: Not sure about this naming `_data_scrap_display` and `_display_scrap_display`? They feel a bit awkward and not quite sure what they do. Perhaps `_prepare_scrap_display` (prepare for data) and `_draw_up_scrap_display` or `_assemble_scrap_display` (draw up or assemble for display).

Comment (MSeal) [reply scrapbook/api.py]: Agreed. I backed off on some of the refactor here and just made the single shared piece accessible. See if the new name makes more sense?

**Comment by MSeal** [comment]
Thanks for the late night review @willingc !

File: scrapbook/models.py [hunk 7]
```diff
@@ -323,23 +281,30 @@ def reglue(self, name, new_name=None, raise_on_missing=True):
...
-                    GLUE_PAYLOAD_FMT.format(encoder=scrap.encoder): scrap_to_payload(
-                        scrap
-                    )
-                }
-                metadata = {"scrapbook": dict(name=name)}
-                # Call raw display function
+                data, metadata = _prepare_ipy_data_format(
+                    scrap.name, scrap_to_payload(scrap), scrap.encoder
+                )
+                # A little non-sensical in this case, but when regluing unattached
```
Comment (willingc) [inline scrapbook/models.py]: Perhaps: Skip saving data for later regluing and remove 'scrapbook' from keys, when unattached

Removing line 293 as well.

File: scrapbook/models.py [hunk 8]
```diff
@@ -323,23 +281,30 @@ def reglue(self, name, new_name=None, raise_on_missing=True):
...
+            if scrap.display is not None:
+                scrap_data = scrap.display.get("data", {})
+                scrap_metadata = self._strip_scrapbook_metadata(
+                    scrap.display.get("metadata", {})
+                )
+                data, metadata = _prepare_ipy_display_format(
+                    scrap.name, scrap_data, scrap_metadata
+                )
+                if unattached:
+                    # Remove 'scrapbook' from keys if we want it unassociated
```
Comment (willingc) [inline scrapbook/models.py]: Perhaps `Remove 'scrapbook' from keys to disassociate display, when unattached

File: scrapbook/models.py [hunk 9]
```diff
@@ -418,9 +385,20 @@ def scraps_report(self, scrap_names=None, notebook_names=None, header=True):
...
         notebook_names : str or iterable[str] (optional)
             notebook names to use in filtering on scraps to report
+        include_data : bool (default: False)
+            indicator that data-only scraps should be reported
         header : bool (default: True)
             indicator for if the scraps should render with a header
         """
+
+        def trim_repr(data):
+            # Used to generate a smallish version of data for display purposes
```
Comment (willingc) [inline scrapbook/models.py]: Convert to one line docstring `"""Generate a small data representation for display purposes"""`

File: setup.py [hunk 10]
```diff
@@ -41,7 +41,8 @@ def read_reqs(fname):
 
 
 reqs_s3 = ["papermill[s3]"]
-reqs_azure = ["papermill[azure]"]
+# Remove requests >= 2.21.0 when papermill releases 0.18.3
```
Comment (willingc) [inline setup.py]: Nice!

**Comment by betatim** [comment]
Merge away.

**Pull request merged**